Add Nautobot App migration docs for Region and Site to Location data migration (#3350)
nautobot/nautobot: nautobot/docs/installation/region-and-site-data-migration-guide.md

# Region and Site to Location Migration Guide

In Nautobot 2.0.0, all the `Region` and `Site` related data models are being migrated to use `Location`. Below is a comprehensive guide for Nautobot users and Nautobot App developers to migrate their `Region` and `Site` related models and apps to `Location`.

## Migrate ObjectPermission instances in Nautobot from Region and Site to Location

> Nautobot admins need to migrate `Site` and `Reigon` Related `ObjectPermission` instances to `Location`. The correct way to do it are documented below with practical examples.

Reviewer: ```suggestion
Nautobot admins need to migrate `Site` and `Region` Related `ObjectPermission` instances to `Location`. The correct way to do it are documented below with practical examples.
```

### Region Specific ObjectPermission

> Add `Location` ContentType to `object_types` field of the `ObjectPermission` if it is not already present.

Reviewer: Here and below, is "add" correct, or should the user "replace" the Region `object_types` entry with a Location entry?

> Since `Location` has all the filters that `Region` had and they retain the same functionalities. We do not need to modify the `constraints` field of any `Region` specific `ObjectPermission` instances.

Reviewer: Might be good to talk about adding a `location_type` constraint to the permission, e.g. if the old permission was one that allowed for editing Regions but not editing Sites? Same comment on the next section as well.

### Site Specific ObjectPermission

Add `Location` ContentType to `object_types` field of the `ObjectPermission` if it is not already present.

The old `constraints` field for `Site` specifc `ObjectPermission` instances might look like this:

```json
{
    "name": "AMS01",
    "name__in": ["AMS01", "ATL01", "EDX01"],
    "slug": "ams01",
    "slug__in": ["ams01", "atl01", "edx01"],
    "id": "9121jfjnjas9fc2-0326-40db-b3a3-3af1231ad",

>     "id__in": ["9121jfjnjas9fc2-0326-40db-b3a3-3afe2207c1ad", "mklmjb79012mkjasd-skmk198721-091231"],

Reviewer: Ok, these definitely aren't real/valid UUIDs. :-) Can you put some "real" UUIDs in here? I got distracted by seeing these nonsense strings.

    "region__name": "United States",
    "region__name__in": ["United States", "Greece", "England"],
    "region__slug": "united-states",
    "region__slug__in": ["united-states", "greece", "england"],
    "region__id": "9e0e9fc2-0326-40db-b3a3-3afe2207c1ad",
    "region__id__in": ["9e0e9fc2-0326-40db-b3a3-3afe2207c1ad", "9e0e9fc2-0326-40db-2222-222222222222"],
    "region__parent__name": "North America",
    "region__parent__name__in": ["North America", "Europe"],
    "region__parent__slug": "north-america",
    "region__parent__slug__in": ["north-america", "europe"],
    "region__parent__id": "9mkl123094fc2-0326-40db-b3a3-3afe2207c1ad",

>     "region__parent__id__in": ["9mkl123094fc2-0326-40db-b3a3-3afe2207c1ad", "82132yhks-875njjni121-0091213"],

Reviewer: Here too, fake UUIDs. :-)

}
```

To modify the data correctly, we need to replace query filter strings to those of model class `Location`:
    1. Replace all occurrences of "region" with "parent" in the **Key** portion (before ":") of the data **not Value** (after ":").

>     2. Since we are keeping the uuids of the newly created "Site"/"Region" type locations the same from those of the old `Site`/`Region` instances, we do not need to change the uuid values in "..__id" and "..__id__in" Keys.

Reviewer: ```suggestion
    2. Since Nautobot carries over the UUIDs of the old `Site`/`Region` instances when creating the new "Site"/"Region" type `Location` instances, we do not need to change the UUID values in `...__id` and `...__id__in` Keys.
```

    "parent__parent__slug__in": ["north-america", "europe"],
    "parent__parent__id": "9mkl123094fc2-0326-40db-b3a3-3afe2207c1ad",
    "parent__parent__id__in": ["9mkl123094fc2-0326-40db-b3a3-3afe2207c1ad", "82132yhks-875njjni121-0091213"],
}
```

### Other Data Model Specific ObjectPermission e.g. Interface

The old `constraints` field for a `Site`/`Region` related data model's (e.g. `Interface`) `ObjectPermission` instance might look like this:

```json
{
    "device__site__name": "AMS01",
    "device__site__name__in": ["AMS01", "ATL01", "ETX02"],
    "device__site__slug": "ams01",
    "device__site__slug__in": ["ams01", "atl01", "etx02"],
    "device__site__id": "0ab47314-2944-45f6-b964-9e009fc48ce0",
    "device__site__id__in": ["0ab47314-2944-45f6-b964-9e009fc48ce0", "b09545d4-6e2b-471e-8f07-27f25ca308f5"],
    "device__site__region__name": "United States",
    "device__site__region__name__in": ["United States", "United Kingdom", "Greece"],
    "device__site__region__slug": "united-states",
    "device__site__region__slug__in": ["united-states", "united-kingdom", "greece"],
    "device__site__region__id": "0ab47314-123144-2132f6-b964-9ekmwkq12c48ce0",

>     "device__site__region__id__in": ["0ab47314-123144-2132f6-b964-9ekmwkq12c48ce0", "kml21345d4-6e123-471e-8as1231-27f25lmsahca308f5"],

Reviewer: More fake UUIDs! Where will the deception end?! ;-)

    "device__site__region__parent__name": "North America",
    "device__site__region__parent__name__in": ["North America", "Europe", "South America"],
    "device__site__region__parent__slug": "north-america",
    "device__site__region__parent__slug__in": ["north-america", "europe", "south-america"],
    "device__site__region__parent__id": "0kmskqjneq14-945840-2132f6-b964-9ekmwkq12c48ce0",
    "device__site__region__parent__id__in": ["0kmskqjneq14-945840-2132f6-b964-9ekmwkq12c48ce0", "qwe0kmjn1lohg-12olm-kmlie-8as1231-km2l1321isdan234f5"],
}
```

To modify the data correctly, we need to replace query filter strings to those of model class `Location`:
    1. Replace all occurrences of "site" with "location" in the **Key** portion (before ":") of the data **not Value** (after ":").
    2. Replace all occurrences of "region" with "parent" in the **Key** portion (before ":") of the data **not Value** (after ":").

>     3. Since we are keeping the uuids of the newly created "Site"/"Region" type locations the same from those of the old `Site`/`Region` instances, we do not need to change the uuid values in "..__id" and "..__id__in" Keys.

Reviewer: ```suggestion
    3. Since Nautobot carries over the UUIDs of the old `Site`/`Region` instances when creating the new "Site"/"Region" type `Location` instances, we do not need to change the UUID values in `...__id` and `...__id__in` Keys.
```

    "device__location__parent__name": "United States",
    "device__location__parent__name__in": ["United States", "United Kingdom", "Greece"],
    "device__location__parent__slug": "united-states",
    "device__location__parent__slug__in": ["united-states", "united-kingdom", "greece"],
    "device__location__parent__id": "0ab47314-123144-2132f6-b964-9ekmwkq12c48ce0",
    "device__location__parent__id__in": ["0ab47314-123144-2132f6-b964-9ekmwkq12c48ce0", "kml21345d4-6e123-471e-8as1231-27f25lmsahca308f5"],
    "device__location__parent__parent__name": "North America",
    "device__location__parent__parent__name__in": ["North America", "Europe", "South America"],
    "device__location__parent__parent__slug": "north-america",
    "device__location__parent__parent__slug__in": ["north-america", "europe", "south-america"],
    "device__location__parent__parent__id": "0kmskqjneq14-945840-2132f6-b964-9ekmwkq12c48ce0",
    "device__location__parent__parent__id__in": ["0kmskqjneq14-945840-2132f6-b964-9ekmwkq12c48ce0", "qwe0kmjn1lohg-12olm-kmlie-8as1231-km2l1321isdan234f5"],
}
```

> ### Other Data Model Specific ObjectPermission e.g. ObjectChange

Reviewer: Use of ObjectChange as a permissions example here feels a bit contrived to me. If we're looking for one with a content-type reference and/or a generic foreign key, maybe something like ExportTemplate might be a better example?

Author: It seems like `owner_content_type` and `owner_object_id` only involve "extras.gitrepository" and `GitRepository` uuids. I might opt for `Note` model as an example.

The old `constraints` field for a `Site`/`Region` related data model's (e.g. `ObjectChange`) `ObjectPermission` instance might look like this:

```json
{
    "changed_object_type": "dcim.site",
    "changed_object_id": "0ab47314-2944-45f6-b964-9e009fc48ce0",
    "changed_object_id__in": ["0ab47314-2944-45f6-b964-9e009fc48ce0", "b09545d4-6e2b-471e-8f07-27f25ca308f5"],
}
```

To modify the data correctly, we need to replace query filter strings to those of model class `Location`:
    1. Replace all occurrences of "dcim.site" with "dcim.location" in the **Value** portion (after ":").

>     3. Since we are keeping the uuids of the newly created "Site"/"Region" type locations the same from those of the old `Site`/`Region` instances, we do not need to change the uuid values in "changed_object_id" and "changed_object_id__in" Keys.

Reviewer: ```suggestion
    3. Since Nautobot carries over the UUIDs of the old `Site`/`Region` instances when creating the new "Site"/"Region" type `Location` instances, we do not need to change the UUID values in the `changed_object_id` and `changed_object_id__in` Keys.
```

## Migrate ObjectPermission instances in Nautobot from Region and Site to Location

Nautobot admins need to migrate `Site` and `Region` Related `ObjectPermission` instances to `Location`. The correct way to do it are documented below with practical examples.

### Region Specific ObjectPermission

Replace `Location` ContentType to `object_types` field of the `ObjectPermission` if it is not already present and add `"location_type__name": "Region"` to `constraints` field if the old `ObjectPermission` only allows operations on `Region` instances and not on `Site` instances.

Since `Location` has all the filters that `Region` had and they retain the same functionalities. We **do not** need to modify the `constraints` field of any `Region` specific `ObjectPermission` instances any further.

### Site Specific ObjectPermission

Replace `Location` ContentType to `object_types` field of the `ObjectPermission` if it is not already present and add `location_type__name: "Site"` to `constraints` field if the old `ObjectPermission` only allows operations on `Site` instances and not on `Region` instances.

> The old `constraints` field for `Site` specifc `ObjectPermission` instances might look like this:

Reviewer: ```suggestion
The old `constraints` field for a `Site` specific `ObjectPermission` instance might look like this:
```

```json
{
    "name": "AMS01",
    "name__in": ["AMS01", "ATL01", "EDX01"],
    "slug": "ams01",
    "slug__in": ["ams01", "atl01", "edx01"],
    "id": "c12753e5-8f01-49a6-b0cf-bf8b460853a1",
    "id__in": ["c12753e5-8f01-49a6-b0cf-bf8b460853a1", "455038c3-4045-4b78-85f5-17d9f34cb9e8"],
    "region__name": "United States",
    "region__name__in": ["United States", "Greece", "England"],
    "region__slug": "united-states",
    "region__slug__in": ["united-states", "greece", "england"],
    "region__id": "f77f5706-e5b3-49e0-9749-b8f818319c40",
    "region__id__in": ["f77f5706-e5b3-49e0-9749-b8f818319c40", "7df99dc2-283a-4130-8125-60b9ca293131"],
    "region__parent__name": "North America",
    "region__parent__name__in": ["North America", "Europe"],
    "region__parent__slug": "north-america",
    "region__parent__slug__in": ["north-america", "europe"],
    "region__parent__id": "c1a816df-876f-44d4-8ea0-335898998780",

>     "region__parent__id__in": ["c1a816df-876f-44d4-8ea0-335898998780", "a68b0838-d7fb-416c-b4ba-a3e464e552ba"],

Reviewer: ```suggestion
    "region__parent__id__in": ["c1a816df-876f-44d4-8ea0-335898998780", "a68b0838-d7fb-416c-b4ba-a3e464e552ba"]
```

The updated JSON data might look like this:

```json
{
    "location_type__name": "Site",
    "name": "AMS01",
    "name__in": ["AMS01", "ATL01", "EDX01"],
    "slug": "ams01",
    "slug__in": ["ams01", "atl01", "edx01"],
    "id": "c12753e5-8f01-49a6-b0cf-bf8b460853a1",
    "id__in": ["c12753e5-8f01-49a6-b0cf-bf8b460853a1", "455038c3-4045-4b78-85f5-17d9f34cb9e8"],
    "parent__name": "United States",
    "parent__name__in": ["United States", "Greece", "England"],
    "parent__slug": "united-states",
    "parent__slug__in": ["united-states", "greece", "england"],
    "parent__id":  "f77f5706-e5b3-49e0-9749-b8f818319c40",
    "parent__id__in": ["f77f5706-e5b3-49e0-9749-b8f818319c40", "7df99dc2-283a-4130-8125-60b9ca293131"],
    "parent__parent__name": "North America",
    "parent__parent__name__in": ["North America", "Europe"],
    "parent__parent__slug": "north-america",
    "parent__parent__slug__in": ["north-america", "europe"],
    "parent__parent__id": "c1a816df-876f-44d4-8ea0-335898998780",

>     "parent__parent__id__in": ["c1a816df-876f-44d4-8ea0-335898998780", "a68b0838-d7fb-416c-b4ba-a3e464e552ba"],

Reviewer: ```suggestion
    "parent__parent__id__in": ["c1a816df-876f-44d4-8ea0-335898998780", "a68b0838-d7fb-416c-b4ba-a3e464e552ba"]
```

    "device__site__name__in": ["AMS01", "ATL01", "ETX02"],
    "device__site__slug": "ams01",
    "device__site__slug__in": ["ams01", "atl01", "etx02"],
    "device__site__id": "0ab47314-2944-45f6-b964-9e009fc48ce0",
    "device__site__id__in": ["0ab47314-2944-45f6-b964-9e009fc48ce0", "b09545d4-6e2b-471e-8f07-27f25ca308f5"],
    "device__site__region__name": "United States",
    "device__site__region__name__in": ["United States", "United Kingdom", "Greece"],
    "device__site__region__slug": "united-states",
    "device__site__region__slug__in": ["united-states", "united-kingdom", "greece"],
    "device__site__region__id": "f1a79a3c-d980-40e1-979d-abdb0f83388e",
    "device__site__region__id__in": ["f1a79a3c-d980-40e1-979d-abdb0f83388e", "6335a61e-503d-463c-99c2-9c87ef8354d9"],
    "device__site__region__parent__name": "North America",
    "device__site__region__parent__name__in": ["North America", "Europe", "South America"],
    "device__site__region__parent__slug": "north-america",
    "device__site__region__parent__slug__in": ["north-america", "europe", "south-america"],
    "device__site__region__parent__id": "6695809c-b33b-4f12-b0de-a4969000434d",

>     "device__site__region__parent__id__in": ["6695809c-b33b-4f12-b0de-a4969000434d", "e51d07bb-3fcf-4306-9d87-6b1ff6dd6378"],

Reviewer: ```suggestion
    "device__site__region__parent__id__in": ["6695809c-b33b-4f12-b0de-a4969000434d", "e51d07bb-3fcf-4306-9d87-6b1ff6dd6378"]
```

    "device__location__slug": "ams01",
    "device__location__slug__in": ["ams01", "atl01", "etx02"],
    "device__location__id": "0ab47314-2944-45f6-b964-9e009fc48ce0",
    "device__location__id__in": ["0ab47314-2944-45f6-b964-9e009fc48ce0", "b09545d4-6e2b-471e-8f07-27f25ca308f5"],
    "device__location__parent__name": "United States",
    "device__location__parent__name__in": ["United States", "United Kingdom", "Greece"],
    "device__location__parent__slug": "united-states",
    "device__location__parent__slug__in": ["united-states", "united-kingdom", "greece"],
    "device__location__parent__id": "f1a79a3c-d980-40e1-979d-abdb0f83388e",
    "device__location__parent__id__in": ["f1a79a3c-d980-40e1-979d-abdb0f83388e", "6335a61e-503d-463c-99c2-9c87ef8354d9"],
    "device__location__parent__parent__name": "North America",
    "device__location__parent__parent__name__in": ["North America", "Europe", "South America"],
    "device__location__parent__parent__slug": "north-america",
    "device__location__parent__parent__slug__in": ["north-america", "europe", "south-america"],
    "device__location__parent__parent__id": "6695809c-b33b-4f12-b0de-a4969000434d",

>     "device__location__parent__parent__id__in": ["6695809c-b33b-4f12-b0de-a4969000434d", "e51d07bb-3fcf-4306-9d87-6b1ff6dd6378"],

Reviewer: ```suggestion
    "device__location__parent__parent__id__in": ["6695809c-b33b-4f12-b0de-a4969000434d", "e51d07bb-3fcf-4306-9d87-6b1ff6dd6378"]
```

}
```

### Other Data Model Specific ObjectPermission e.g. Note

The old `constraints` field for a `Site`/`Region` related data model's (e.g. `Note`) `ObjectPermission` instance might look like this:

```json
{
    "assigned_object_type": "dcim.site",
    "assigned_object_id": "932d94ee-5571-40a0-903f-4274fcfbed32",

>     "assigned_object_id__in": ["932d94ee-5571-40a0-903f-4274fcfbed32", "e383db9a-dd55-464d-9e56-2f18bc03b32c"],

Reviewer: ```suggestion
    "assigned_object_id__in": ["932d94ee-5571-40a0-903f-4274fcfbed32", "e383db9a-dd55-464d-9e56-2f18bc03b32c"]
```

}
```

To modify the data correctly, we need to:
    1. Replace all occurrences of "dcim.site" and/or "dcim.region" with "dcim.location" in the **Value** portion (after ":") of the `assigned_object_type` Key.
    2. Since Nautobot carries over the UUIDs of the old `Site`/`Region` instances when creating the new "Site"/"Region" type `Location` instances, we **do not** need to change the UUID values in the `assigned_object_id` and `assigned_object_id__in` Keys.

The updated JSON data might look like this:

```json
{
    "assigned_object_type": "dcim.location",
    "assigned_object_id": "932d94ee-5571-40a0-903f-4274fcfbed32",

>     "assigned_object_id__in": ["932d94ee-5571-40a0-903f-4274fcfbed32", "e383db9a-dd55-464d-9e56-2f18bc03b32c"],

Reviewer: ```suggestion
    "assigned_object_id__in": ["932d94ee-5571-40a0-903f-4274fcfbed32", "e383db9a-dd55-464d-9e56-2f18bc03b32c"]
```

### Add Location Fields to Site and Region Related Data Models

If the `ExampleModel` currently has a `site` ForeignKey field but it does not have a `location` ForeignKey field, you will need to add the `location` field before any other migrations in this guide.

!!! note
    You can skip this step only if your data models already have both a `site` (or `region`) field and a `location` field.

```python
# models.py

class ExampleModel(OrganizationalModel):
    site = models.ForeignKey(to="dcim.Site", on_delete=models.PROTECT, blank=True, null=True)
    name = models.CharField(max_length=20, help_text="The name of this Example.")
...
```

**DO NOT** delete the `site` ForeignKey field yet. As a first step, just add a `ForeignKey` to `dcim.Location` with all other arguments identical to the existing `dcim.Site` `ForeignKey`:

```python
# models.py

class ExampleModel(OrganizationalModel):
    site = models.ForeignKey(to="dcim.Site", on_delete=models.PROTECT, blank=True, null=True)
    location = models.ForeignKey(to="dcim.Location", on_delete=models.PROTECT, blank=True, null=True)
    name = models.CharField(max_length=20, help_text="The name of this Example.")
...
```

> Make the migration file by running `nautobot-server makemigrations [app_name] -n [migration_name]`, for example:

Reviewer: What state do the Nautobot migrations need to be in at this point? Still a 1.5.x instance, or should 2.0 be already installed (but not yet migrated) at this point?

...
```

After removing the `site` attribute, make the migration file by running `nautobot-server makemigrations [app_name] -n [migration_name]`, for example:

```shell
nautobot-server makemigrations example_app -n remove_site_field_from_example_model
```

The migration file might look like this:

```python
# Generated by Django 3.2.17 on 2023-02-22 17:09
# 0009_remove_site_field_from_example_model.py

from django.db import migrations

class Migration(migrations.Migration):

    dependencies = [
        ('example_app', '0008_migrate_example_model_data_from_site_to_location'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='examplemodel',
            name='site',
        ),
    ]
```

!!! important
    Before you apply this migration, you have to add a `run_before` attribute in this migration file to make sure that you remove `site`/`region` fields before `Site` and `Region` models themselves are removed. **Without it, your migration files might be out of order and your app will not start**.

```python
    # Ensure this migration is run before the migration that removes Region and Site Models
    run_before = [

>         # TODO we need to change the name after we reorder migration files

Reviewer: When will this be done?

Author: After we merge #3313 I think, added that to the TODO comment.

        ("dcim", "0034_remove_region_and_site"),
    ]
```

The final migration file might look like this:

```python
# Generated by Django 3.2.17 on 2023-02-22 17:09
# 0009_remove_site_field_from_example_model.py

from django.db import migrations

class Migration(migrations.Migration):
    
    # Ensure this migration is run before the migration that removes Region and Site Models
    run_before = [
        # TODO we need to change the name after we reorder migration files
        ("dcim", "0034_remove_region_and_site"),
    ]
    dependencies = [
        ("example_app", "0008_migrate_example_model_data_from_site_to_location"),
    ]

    operations = [
        migrations.RemoveField(
            model_name="examplemodel",
            name="site",
        ),
    ]
```

!!!important
    Before you apply the aforementioned migration files to your app in the following step, please make sure that Nautobot is on version 1.x and not on version 2.0. More importantly, please make sure that `("dcim", "0030_migrate_region_and_site_data_to_locations")` is applied to your Nautobot instance and that `("dcim", "0034_remove_region_and_site")` is **not** applied.

> 

Reviewer: Would be good to flesh this out a bit further with exact steps, something like:

- Start with a working Nautobot 1.5.x installation with baseline migrations all run
- Stop Nautobot server
- Update installed version of Nautobot to 2.0.0
- Run `nautobot-server migrate dcim 0030_migrate_region_and_site_data_to_locations` (?)
- Updated installed version of *all* Apps to include any necessary migration files as described above
- For each affected app, run `nautobot-server migrate <app_name>` (?)
- Run `nautobot-server migrate` to finish all migrations (?)
- Start Nautobot server

Author: I moved this to before the guide starts

    1. Replace all occurrences of "dcim.site" and/or "dcim.region" with "dcim.location" in the **Value** portion (after ":") of the `assigned_object_type` Key.
    2. Since Nautobot carries over the UUIDs of the old `Site`/`Region` instances when creating the new "Site"/"Region" type `Location` instances, we **do not** need to change the UUID values in the `assigned_object_id` and `assigned_object_id__in` Keys.

The updated JSON data might look like this:

```json
{
    "assigned_object_type": "dcim.location",
    "assigned_object_id": "932d94ee-5571-40a0-903f-4274fcfbed32",
    "assigned_object_id__in": ["932d94ee-5571-40a0-903f-4274fcfbed32", "e383db9a-dd55-464d-9e56-2f18bc03b32c"]
}
```

## Region and Site Related Data Model Migration Guide For Existing Nautobot App installations

In Nautobot 2.0.0, all the `Region` and `Site` related data models are being migrated to use `Location`. Below is a comprehensive guide for Nautobot App developers to migrate their `Region` and `Site` related data models to `Location`.

We will be using `ExampleModel` as a relatively simple and hands-on example throughout this guide to better your understanding of the migration process.

> !!!important  
>     Before you follow the guide, please make sure that these operations are completed:  
>         1. Make sure your working Nautobot is on Version 1.5.x with baseline migrations all run.  
>         2. Stop Nautobot Server.  
>         3. Update installed version of Nautobot to 2.0.0.  
>         4. Run `nautobot-server migrate dcim 0030_migrate_region_and_site_data_to_locations`. (This step will ensure that `("dcim", "0030_migrate_region_and_site_data_to_locations")` is applied to your Nautobot instance and that `("dcim", "0034_remove_region_and_site")` is **not** applied.)  
>     After you complete those operations, follow the guide below for each of your installed apps to:  
>         1. Make all necessary migration files for each app.  
>         2. Run `nautobot-server migrate [app_name]` to apply those migration files to each app.  
>         3. Finally, Start Nautobot Server after **all** the migration files are applied and **all** your affected apps are updated.  

Reviewer 2: This is a lot of information stuffed into an admonition.

It would be better as an h3 section with an admonition at the top that is more than just "important". Probably "warning" so that it's orange and stands out. 

It would be good for this admonition to say something like "You must perform these steps." It needs to be very explicit that if you just run the migrations without stopping at `0030` that you absolutely will nuke your data.

Additionally, there should be an extra step after `2. Stop Nautobot Server` that tells the reader to make a database backup. We don't have any documentation on how to backup data. The closest thing we have is a section on "Export data from PostgreSQL" in the "Migrating from PostgreSQL to MySQL" document.

The suggested changes are not complete. Just to give an example.

```suggestion
### Before you Begin 

!!! warning
    You **must** perform these steps before proceeding. Failing to follow them properly could result in data loss. Always backup your database before performing any migration steps.
    
Before you follow the guide, please make sure that these operations are completed:  

1. Make sure your working Nautobot is on Version 1.5.x with baseline migrations all run.  
2. Stop Nautobot Server.  
3. Create a backup of your Nautobot database.
4. Update installed version of Nautobot to 2.0.0.  
5. Run `nautobot-server migrate dcim 0030_migrate_region_and_site_data_to_locations`. (This step will ensure that `("dcim", "0030_migrate_region_and_site_data_to_locations")` is applied to your Nautobot instance and that `("dcim", "0034_remove_region_and_site")` is **not** applied.)  

After you complete those operations, follow the guide below for each of your installed apps to:  

1. Make all necessary migration files for each app.  
2. Run `nautobot-server migrate [app_name]` to apply those migration files to each app.  
3. Finally, Start Nautobot Server after **all** the migration files are applied and **all** your affected apps are updated.  
```

Author: Is `nautobot-server dumpdata` something you can use to back up your data?

Reviewer 2: Short answer is yes. It is a solution that works across any database backend.

```python
    # Ensure this migration is run before the migration that removes Region and Site Models
    run_before = [
        # TODO we need to change the name when PR #3313 is merged and migrations files are reordered.
        ("dcim", "0034_remove_region_and_site"),
    ]
```

The final migration file might look like this:

```python
# Generated by Django 3.2.17 on 2023-02-22 17:09
# 0009_remove_site_field_from_example_model.py

from django.db import migrations

class Migration(migrations.Migration):
    
    # Ensure this migration is run before the migration that removes Region and Site Models
    run_before = [
        # TODO we need to change the name when PR #3313 is merged and migrations files are reordered.
        ("dcim", "0034_remove_region_and_site"),
    ]
    dependencies = [
        ("example_app", "0008_migrate_example_model_data_from_site_to_location"),
    ]

    operations = [
        migrations.RemoveField(
            model_name="examplemodel",
            name="site",
        ),
    ]
```

Apply the migration files by running `nautobot-server migrate [app_name]`, for example:

```shell
nautobot-server migrate example_app
```

> ## Region and Site Related Data Model Migration Guide For New Nautobot App installations in an Existing Nautobot 2.0 Environment

Reviewer 2: Is this still a work-in-progress? Why an empty section?

Author: I need to get an app that has long-running migrations and that uses `Site` and `Region` models in it. So I can try to install it in our alpha environment to know what exactly to do when installing new apps in the 2.0.0 environment.

Author: I didn't want to hold up this PR anymore so I thought I could just merge it and add on to it later.

Reviewer 2: It's fine then, if you wouldn't mind submitting that on the backlog so we don't lose track of it if it isn't already?

Author: https://github.com/nautobot/nautobot/issues/3387

We will be using `ExampleModel` as a relatively simple and hands-on example throughout this guide to better your understanding of the migration process.

### Before you Begin 
!!! warning
    You **must** perform these steps before proceeding. Failing to follow them properly could result in data loss. **Always** backup your database before performing any migration steps.
    
Before you follow the guide, please make sure that these operations are completed:  
1. Make sure your working Nautobot is on Version 1.5.x with baseline migrations all run.  
2. Stop Nautobot Server.  

> 3. Create a backup of your Nautobot database. (Check out the [Export Data Guide](../installation/migrating-from-postgresql.md#export-data-from-postgresql) for this operation)

Reviewer 2: This is great! My only concern here, and I think it should be done in a separate story is this:

The document that this links to is explicitly under the heading “Export data from PostgreSQL”, so I’m thinking what if readers are using MySQL and see this, will they get stuck because it says PG?

We should revise the migration docs to have another section for generic “dump/restore” Nautobot that isn’t necessarily specific to a backend. OR if it is specific to a backend, it has instructions for each (e.g. using `pgdump` vs. `mysqldump` ).

Author: https://github.com/nautobot/nautobot/issues/3386
